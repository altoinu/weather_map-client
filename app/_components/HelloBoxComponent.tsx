"use client";

import { Button, Stack } from "@mui/material";
import { useState } from "react";

type HomePageProps = {
  personName: string;
  secondPersonName?: string;
  /**
   * Callback fired when the button is clicked.
   *
   * @param {Number} buttonNum Button number clicked.
   */
  onButtonClick?: (buttonNum: number) => void;
};

const HelloBoxComponent = ({
  personName,
  onButtonClick,
  ...props
}: HomePageProps) => {
  const [clickedButtonResponse, setClickedButtonResponse] = useState<string>();

  const handleButtonClick = (buttonNum: number) => {
    const response =
      buttonNum === 1
        ? "First button clicked"
        : buttonNum === 2
          ? "Second button clicked"
          : "something else clicked";

    setClickedButtonResponse(response);

    onButtonClick?.(buttonNum);
  };

  return (
    <>
      <h1>
        Hello {personName}
        {props.secondPersonName ? ` and ${props.secondPersonName}` : ""}!
      </h1>
      {onButtonClick && (
        <Stack direction="row" gap={1}>
          <Button variant="contained" onClick={() => handleButtonClick(1)}>
            Click Me!
          </Button>
          <Button variant="contained" onClick={() => handleButtonClick(2)}>
            Click Me Too!
          </Button>
        </Stack>
      )}
      {clickedButtonResponse && (
        <p>clicked button response: {clickedButtonResponse}</p>
      )}
    </>
  );
};

export default HelloBoxComponent;
