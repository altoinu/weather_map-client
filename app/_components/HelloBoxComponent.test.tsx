import HelloBoxComponent from "./HelloBoxComponent";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("<HelloBoxComponent>", () => {
  it("should render", () => {
    const { container: HomePageContainer } = render(
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={jest.fn}
      />,
    );

    expect(HomePageContainer).toMatchSnapshot();
  });

  it("should have person's name", () => {
    render(<HelloBoxComponent personName="Kaoru" />);
    const heading = screen.getByRole("heading");
    expect(heading.textContent).toBe("Hello Kaoru!");
  });

  it("should have person and second person's names", () => {
    render(<HelloBoxComponent personName="John" secondPersonName="Doe" />);
    const heading = screen.getByRole("heading");
    expect(heading.textContent).toBe("Hello John and Doe!");
  });

  it("should have two buttons when onButtonClick is set", () => {
    render(
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={jest.fn}
      />,
    );

    const button1 = screen.getByRole("button", { name: "Click Me!" });
    expect(button1).toBeInTheDocument();
    const button2 = screen.getByRole("button", { name: "Click Me Too!" });
    expect(button2).toBeInTheDocument();
  });

  it("should contain corresponding clicked button response in body when either button is clicked", async () => {
    const user = userEvent.setup();

    const handleButtonClick = jest.fn();
    render(
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={handleButtonClick}
      />,
    );
    const button1 = screen.getByRole("button", { name: "Click Me!" });
    const button2 = screen.getByRole("button", { name: "Click Me Too!" });

    await user.click(button1);

    expect(
      screen.getByText("clicked button response: First button clicked"),
    ).toBeInTheDocument();

    await user.click(button2);

    expect(
      screen.getByText("clicked button response: Second button clicked"),
    ).toBeInTheDocument();
  });

  it("should call onButtonClick with 'First button clicked' when first button is clicked", async () => {
    const user = userEvent.setup();

    const handleButtonClick = jest.fn();
    render(
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={handleButtonClick}
      />,
    );
    const button1 = screen.getByRole("button", { name: "Click Me!" });

    await user.click(button1);

    expect(handleButtonClick).toHaveBeenCalledWith(1);
  });

  it("should call onButtonClick with 'Second button clicked' when first button is clicked", async () => {
    const user = userEvent.setup();

    const handleButtonClick = jest.fn();
    render(
      <HelloBoxComponent
        personName="John"
        secondPersonName="Doe"
        onButtonClick={handleButtonClick}
      />,
    );
    const button2 = screen.getByRole("button", { name: "Click Me Too!" });

    await user.click(button2);

    expect(handleButtonClick).toHaveBeenCalledWith(2);
  });
});
