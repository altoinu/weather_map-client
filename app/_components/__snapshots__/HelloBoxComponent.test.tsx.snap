// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelloBoxComponent> should render 1`] = `
<div>
  <h1>
    Hello 
    John
     and Doe
    !
  </h1>
  <div
    class="MuiStack-root css-1oxswqj-MuiStack-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Click Me!
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1abvrni-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Click Me Too!
    </button>
  </div>
</div>
`;
